### PR TITLE
[#242] Added error stream to daru

### PR DESCRIPTION
**Pull request by sivagollapalli**
#242

### CODE REVIEW

**Comment by sivagollapalli** [comment]
@zverok I thought that I have to use on https://github.com/SciRuby/daru/blob/master/lib/daru.rb#L87 to silence that warning, but since it is an installation error I thought it would be better to show. And also on remaining code base I didn't see any `STDERR.puts` statements. So I didn't make any changes. If you want to silence the error message I would really happy to make that change.

**Comment by v0dro** [comment]
Here's how you can proceed:
- Set the error_stream default to `STDERR`.
- Change all calls to `STDERR` in the code base to `@@error_stream`.
- Write tests accordingly.

**Comment by sivagollapalli** [comment]
@v0dro Since I didn't know how to test spreadsheet dependency in spec so I avoided that. Could you give suggestions regarding that?

File: lib/daru.rb [hunk 1]
```diff
@@ -72,6 +74,14 @@ def plotting_library= lib
...
       end
     end
+
+    def error_stream= stream 
+      class_variable_set(:@@error_stream, stream)
+    end
+
+    def error_stream
+      @@error_stream
+    end
```
Comment (zverok) [inline lib/daru.rb]: Why did you decided to go with class variable? Instance variable of `Daru` module would be enough -- with those two methods becoming just `attr_accessor :error_stream`

File: lib/daru.rb [hunk 2]
```diff
@@ -84,8 +94,8 @@ def plotting_library= lib
   gem 'spreadsheet', '~>1.1.1'
   require 'spreadsheet'
 rescue LoadError
-  STDERR.puts "\nInstall the spreadsheet gem version ~>1.1.1 for using"\
-  ' spreadsheet functions.'
+  Daru.error_stream.puts "\nInstall the spreadsheet gem version ~>1.1.1 for using"\
+  ' spreadsheet functions.' if Daru.error_stream
```
Comment (zverok) [inline lib/daru.rb]: I'd abstract this into method like `Daru.error(message)`. Then it would be easy to test.

**Comment by sivagollapalli** [comment]
@zverok Added changes. Could you please check?

File: spec/daru_spec.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    Daru.error('priting error message')
+    expect($stderr).to have_received(:puts)
+  end
+
+  it 'should not print error message if it is set to nil' do
+    Daru.error_stream = nil
+    allow($stderr).to receive(:puts).and_return(true)
+    Daru.error('priting error message')
+    expect($stderr).not_to have_received(:puts)
+  end
```
Comment (zverok) [inline spec/daru_spec.rb]:
```ruby
describe '#error' do
  context 'by default' do
    it { expect { Daru.error('test') }.to output('test').to_stderr }
  end

  context 'when set to nil' do
    before { Daru.error_stream = nil }
    it { expect { Daru.error('test') }.not_to output('test').to_stderr }
  end

  context 'when set to instance of custom class' do
    let(:custom_stream) { double(puts: nil) }
    before { Daru.error_stream = custom_stream }

    it 'calls puts' do
      expect { Daru.error('test') }.not_to output('test').to_stderr
      expect(custom_stream).to have_received(:puts).with('test')
    end
  end
end
```
Or something like that.

File: spec/daru_spec.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+describe 'Daru' do
+  it 'default error stream should be $stderr' do
+    expect(Daru.error_stream).to eq($stderr)
+  end
+
+  it 'should be able to set error stream to nil' do
+    Daru.error_stream = nil
+    expect(Daru.error_stream).to be_nil
+  end
```
Comment (zverok) [inline spec/daru_spec.rb]:
```ruby
it { expect { Daru.error_stream = nil }.to change(Daru, :error_stream).to(nil) }
```
Though, this test is completely unnecessary, as you are, in fact, testing Ruby's `attr_accessor` implementation (and the fact of existance of `error_stream=` is tested below anyways).

**Comment by sivagollapalli** [comment]
@zverok Incorporated the changes. Could you please check?

**Pull request merged**